### PR TITLE
Add waypoint precision

### PR DESCRIPTION
**Pull request by appgurueu**
# Overview

- Goal of the PR: Add waypoint precision ("digits-after-dot"). If set to 0, distance is not shown.
- How does the PR work? It uses the `item` field to store precision.
- If not a bug fix, why is this PR needed? What usecases does it solve? Closes [this feature request](https://github.com/minetest/minetest/issues/8141).

## To do

This PR is Ready for Review.

## How to test

Create a mod with the following `init.lua`:
```lua
minetest.register_chatcommand(
	"test_waypoints",
	{
		description = "tests waypoints",
		func = function(sendername, params)
			local player = minetest.get_player_by_name(sendername)
			minetest.chat_send_all(
				"regular waypoint: " ..
					player:hud_add(
						{
							hud_elem_type = "waypoint",
							name = "origin",
							text = "",
							number = 0xFF0000,
							world_pos = vector.add(player:get_pos(), {x = 0, y = 2, z = 0})
						}
					)
			)
			minetest.chat_send_all(
				"better waypoint: " ..
					player:hud_add(
						{
							hud_elem_type = "waypoint",
							name = "origin",
							text = "",
							item = 5,
							number = 0xFF0000,
							world_pos = player:get_pos()
						}
					)
			)
			minetest.chat_send_all(
				"better waypoint without distance: " ..
					player:hud_add(
						{
							hud_elem_type = "waypoint",
							name = "origin",
							text = "",
							item = 0,
							number = 0xFF0000,
							world_pos = player:get_pos()
						}
					)
			)
		end
	}
)
```

## Screenshot

Screenshot

### CODE REVIEW

File: src/script/common/c_content.cpp [hunk 1]
```diff
@@ -1852,6 +1852,9 @@ void read_hud_element(lua_State *L, HudElement *elem)
 	elem->text    = getstringfield_default(L, 2, "text", "");
 	elem->number  = getintfield_default(L, 2, "number", 0);
 	elem->item    = getintfield_default(L, 2, "item", 0);
+	if (elem->type == HUD_ELEM_WAYPOINT) {
```
Comment (appgurueu) [inline src/script/common/c_content.cpp]: Probably this should be changed to an if-else?

File: src/script/common/c_content.cpp [hunk 2]
```diff
@@ -1852,6 +1852,9 @@ void read_hud_element(lua_State *L, HudElement *elem)
...
+		elem->item = getintfield_default(L, 2, "item", -1) + 1;
+	}
```
Comment (appgurueu) [inline src/script/common/c_content.cpp]: Probably this should be changed to an if-else?

Comment (MoNTE48) [reply src/script/common/c_content.cpp]: it seems that yes

Comment (appgurueu) [reply src/script/common/c_content.cpp]: Done

File: src/script/common/c_content.cpp [hunk 3]
```diff
@@ -1851,9 +1851,10 @@ void read_hud_element(lua_State *L, HudElement *elem)
 	elem->name    = getstringfield_default(L, 2, "name", "");
 	elem->text    = getstringfield_default(L, 2, "text", "");
 	elem->number  = getintfield_default(L, 2, "number", 0);
-	elem->item    = getintfield_default(L, 2, "item", 0);
 	if (elem->type == HUD_ELEM_WAYPOINT) {
 		elem->item = getintfield_default(L, 2, "item", -1) + 1;
```
Comment (appgurueu) [inline src/script/common/c_content.cpp]: Should I rename this to `precision`?

**Comment by Wuzzy2** [comment]
> If set to 0, distance is not shown.

But what if I want to show a distance with whole numbers only? Looks like then I'm screwed.

**Comment by Wuzzy2** [comment]
Suggestion: Use `item=-1` for hiding the distance number. Assuming the `item` field is signed …

**Comment by appgurueu** [comment]
@Wuzzy2 unfortunately, it's not: `u32 item;` in HUD.h

File: src/script/common/c_content.cpp (hunk 3 above)
Comment (MoNTE48) [reply src/script/common/c_content.cpp]: and btw, { } are not needed in this case. wWthout them, the code looks neater

Comment (appgurueu) [reply src/script/common/c_content.cpp]: I know, but I prefer to use them.

Comment (appgurueu) [reply src/script/common/c_content.cpp]: Update: Will remove the curly braces according to the preferred code style

**Comment by SmallJoker** [comment]
Why `item`? A field name like `scale` would be clearer for mods, and is versatile. If needed, extend the HUD packet to send an additional `f32`. Old clients won't handle the precision anyway.

**Comment by appgurueu** [comment]
@SmallJoker re-using the currently unused `item` field seems fine to me, as it stores an `u32` which suffices for the precision purpose. `scale` is a vector. And I would prefer not to send an uneeded `f32`.

**Comment by Wuzzy2** [comment]
Nice trick. But I would have been completely fine if the `item` field would be reused. Anyway, it's OK either way.

File: src/script/common/c_content.cpp [hunk 4]
```diff
@@ -1851,7 +1851,10 @@ void read_hud_element(lua_State *L, HudElement *elem)
 	elem->name    = getstringfield_default(L, 2, "name", "");
 	elem->text    = getstringfield_default(L, 2, "text", "");
 	elem->number  = getintfield_default(L, 2, "number", 0);
-	elem->item    = getintfield_default(L, 2, "item", 0);
+	if (elem->type == HUD_ELEM_WAYPOINT)
```
Comment (sfan5) [inline src/script/common/c_content.cpp]: should add a comment here noting that waypoints reuse `item`

File: src/client/hud.cpp [hunk 5]
```diff
@@ -372,11 +370,22 @@ void Hud::drawLuaElements(const v3s16 &camera_offset)
...
-				text = unescape_translate(utf8_to_wide(os.str()));
-				pos.Y += text_height;
-				font->draw(text.c_str(), size + pos, color);
+				// item = precision + 1
+				float precision;
+				u32 item = e->item;
+				if (item == 0)
+					precision = 10.0f;
+				else
+					precision = item - 1;
```
Comment (sfan5) [inline src/client/hud.cpp]: `float precision = e->item == 0 ? 10.0f : (item - 1);`

**Comment by Extex101** [comment]
Would you consider #9447

File: src/script/common/c_content.cpp (hunk 4 above)
Comment (appgurueu) [reply src/script/common/c_content.cpp]: will do

File: src/client/hud.cpp (hunk 5 above)
Comment (appgurueu) [reply src/client/hud.cpp]: will change

File: src/client/hud.cpp [hunk 6]
```diff
@@ -370,13 +370,9 @@ void Hud::drawLuaElements(const v3s16 &camera_offset)
 				core::rect<s32> size(0, 0, 200, 2 * text_height);
 				std::wstring text = unescape_translate(utf8_to_wide(e->name));
 				font->draw(text.c_str(), size + pos, color);
-				// item = precision + 1
-				float precision;
+				// waypoints reusing item for precision, item = precision + 1
```
Comment (sfan5) [inline src/client/hud.cpp]:
```suggestion
				// waypoints reuse item to store the precision, item = precision + 1
```

File: src/script/common/c_content.cpp (hunk 4 above)
Comment (sfan5) [reply src/script/common/c_content.cpp]: This doesn't really explain it, I was thinking more of:
`// Waypoints reuse the item property to store the precision (+1) of the displayed distance`

**Comment by appgurueu** [comment]
Closing this to open a larger PR incorporating it.

**Pull request closed without merging**